ddtrace/tracer/payload.go

```diff
@@ -45,36 +44,38 @@ var _ io.Reader = (*payload)(nil)
 func newPayload() *payload {
 	p := &payload{
 		header: make([]byte, 8),
+		buf:    bytes.NewBuffer(make([]byte, 0, 512)),
 		off:    8,
 	}
 	return p
 }
 
 // push pushes a new item into the stream.
 func (p *payload) push(t spanList) error {
-	if err := msgp.Encode(&p.buf, t); err != nil {
-		return err
-	}
-	p.count++
+	p.traces = append(p.traces, t)
 	p.updateHeader()
 	return nil
 }
 
-// itemCount returns the number of items available in the srteam.
+// itemCount returns the number of items available in the stream.
 func (p *payload) itemCount() int {
-	return int(p.count)
+	return len(p.traces)
 }
 
 // size returns the payload size in bytes. After the first read the value becomes
 // inaccurate by up to 8 bytes.
 func (p *payload) size() int {
-	return p.buf.Len() + len(p.header) - p.off
+	size := len(p.header)
+	for _, t := range p.traces {
+		size += t.Msgsize()
+	}
+	return size
 }
 
 // reset resets the internal buffer, counter and read offset.
 func (p *payload) reset() {
 	p.off = 8
-	p.count = 0
+	p.traces = nil
```

Comment:
Can we simply reuse this?
```go
p.traces = p.traces[:0]
```
We can continue to keep using the underlying array. Assigning `nil` will just get it garbage collected.

Reply:
The case I thought about here is that upon having a traces burst making the payload close to its maximum size, the underlying array can grow significantly, and use up memory even if the given service gets into a "calmer" period in which only a few traces per payload flush are present. Moreover traces still referenced by the underlying array will not be, if I am not mistaken, garbage-collected in this case.

Reply:
Yes, I definitely did consider that. Since we keep track of the relative memory usage by monitoring the (msgpack) payload size and not allowing it to go past 5-10MB max, is this still a concern without considering edge-cases? Is this also a case of premature optimization?

I'm also pondering whether `p.traces = []spanList{}` would provide any benefit over `nil`, but I'd be surprised if it did.

Overall, I tend to lean on the side of my initial suggestion if the speed gain is that big.

Reply:
I just benchmarked this, and it seems like a very nice solution to this edge case. The impact in performance is pretty much non-existent (1-2% on my machine). I added this to p.reset.